### ISSUE TITLE
Autosave slot hierarchy

### ISSUE THREAD

**Issue by 3x1t-5tyl3** [Yellow-Dog-Man/Resonite-Issues#3184]
### Is your feature request related to a problem? Please describe.

When working on items one does tend to forget to save manually *a lot*
We've all done it.

### Describe the solution you'd like

A way for an item to be flagged as "auto-save" (component or otherwise in the inventory. I don't care tbh)

Which would save the hierarchy of a slot downwards.

Give it a slot -> autosave will happen on that slot to a folder automagically

And withotu saving all the other crap laying around in the session one might not want to save from other users; or it being an avatar ect. making it unfeasable in someone elses world obviously.

### Describe alternatives you've considered

N/A

### Requesters

3x1t_5tyl3

**Comment by shiftyscales** [comment]
Can you clarify your reasons for this request, @3x1t-5tyl3? Is it primarily because you're wanting to work on an item in a public space around other users / sessions that are not your own? If so, there may possibly be other / better approaches to this problem, e.g. having a local layer / "bubble session" #1118.

Everything you are working on in a given world is not guaranteed to be cleanly self-contained in a single slot, e.g. if you make new Flux nodes that haven't yet been packed- they'd go under the root- meaning you'd have to save the entire world hierarchy to save everything you were working on- at which point it'd just be better to use the existing auto-save mechanism for worlds instead.

There are other existing components you could use to improve your workflow / create your own tooling, e.g. if you created a "work platform" that uses grabbable parenter to parent anything dropped on top of it, and just periodically grab the work space directly and save it to your inventory in the context menu. This would be a quick and relatively simple way of ensuring you capture everything you want to- it'd still require you to manually save- but you could also build reminders into the work platform to save it periodically / send you reminders- if your creation is finished / not in a WIP state anymore- you could just grab it directly off of the platform and save it .

Another useful tool is the "ExternalSlotSaver" component- when you initialize a save on a slot that contains that component- it will save the TargetSlot specified with whatever else you are saving. Used in combination with the above- you could save your WIP workspace alongside your avatar just by saving your avatar

In general- if you're working on something important to you- I'd recommend you host your own sessions / headless, and make use of existing world auto-save functionality to ensure everything you could want to save is captured- even if some stuff you don't care about is as well.

**Comment by AmasterAmaster** [comment]
This hits the nail on the head on why many people, including myself, prefer to hermit in private sessions (to work on worlds, avatars, or even simple items) and not socialize with others (to work on said items). Solving this is a good step forward to a more inclusive and social community.

**Comment by Nytra** [comment]
Can't you just put a giant piece of text in the world that says "Remember to save!"?

**Comment by ko-tengu** [comment]
Even outside of the bigger picture things, it's a very basic type of automation that the vast majority of users are familiar with. Saving the single object you're working on to a file that is just the object is much more friendly and approachable than saving an entire world even before we get to the various performance and bugs that can make auto-saving the entire world less appealing.

Having an auto-save would also help a lot of peace of mind in sessions that may randomly crash or even just getting distracted for a few minutes by talking to someone at the end of your session and saving slipping your mind. Reminders and floating text and all that only work if you're actively thinking about it to begin with.

**Comment by shiftyscales** [comment]
Hmm- there are some things that'd probably need to be changed / further developed before this could even be considered.

E.g. having some mechanism to designate a particular hierarchy / set of hierarchies as a discrete "object", e.g. something like #417.

As peviously highlighted- there are creations and cases in which a single "object" doesn't necessarily exist exclusively under just a single hierarchy, and particularly once you get into aforementioned cases like newly made Flux that hasn't yet ever been packed into the 'object' but otherwise should still be considered part of it.

Likewise, this proposed functionality would also be greatly enhanced by (and I think would be best handled after) molecule #1212 to have a native versioning system- otherwise, the proposed option would rapidly fill up the users' inventories with partially-completed copies of objects which would all otherwise be named identically- making it difficult to tell which version is which without spawning them out and checking. Having a proper versioning system would help resolve this.

Until then- the best users can do for themselves is to better manage their workflow- save early, save often.

---

Specific to the case of world crashes- when they're managed crashes, Resonite will attempt to dump the world state to disk so the record can be loaded- if this happens, exit Resonite and check your log file to see if the world state was dumped / can be recovered by dragging the dumped record onto the Resonite window during the following session.

**Comment by 3x1t-5tyl3** [comment]
> Hmm- there are some things that'd probably need to be changed / further developed before this could even be considered.
> 
> E.g. having some mechanism to designate a particular hierarchy / set of hierarchies as a discrete "object", e.g. something like [#417](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/417).

Agreed!

> As peviously highlighted- there are creations and cases in which a single "object" doesn't necessarily exist exclusively under just a single hierarchy, and particularly once you get into aforementioned cases like newly made Flux that hasn't yet ever been packed into the 'object' but otherwise should still be considered part of it.

We pack things soemwhat consistently when working on stuff. I don't know of many usecases for not packing and unpacking constantly?
And even if one is saved without it. An incremental upgrade to it is still far better than none.
Ignorign the fact we already have very popular tools (or tools I made myself) that parent themselves underneath the item so that any flux is also instantly parented underneat the item.

> Likewise, this proposed functionality would also be greatly enhanced by (and I think would be best handled after) molecule [#1212](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1212) to have a native versioning system- otherwise, the proposed option would rapidly fill up the users' inventories with partially-completed copies of objects which would all otherwise be named identically- making it difficult to tell which version is which without spawning them out and checking. Having a proper versioning system would help resolve this.
> 
> Until then- the best users can do for themselves is to better manage their workflow- save early, save often.

I somewhat agree. It would be nice; However we do have very easy ways to make our own versioning system. So it might not be the single-most blocking thing that would entirely need to be done before this. Though I will agree that it would help heaps!

To be quite honest; I'd rather have 100 of a similar item before I lose another insane amount of work due to me getting distracted thanks to things that are outside my control (such as ADHD, phonecalls, emails I need to write ect)

> Specific to the case of world crashes- when they're managed crashes, Resonite will attempt to dump the world state to disk so the record can be loaded- if this happens, exit Resonite and check your log file to see if the world state was dumped / can be recovered by dragging the dumped record onto the Resonite window during the following session.

**Comment by 3x1t-5tyl3** [comment]
> Can't you just put a giant piece of text in the world that says "Remember to save!"?

Yeah no. This is unhelpful here. If you read the post: The issue is not in what I save. But how it is saved. Any tool to help you remember to save is *always* going to be subpar compared to auto-saves.

**Comment by Nytra** [comment]
You literally said

> **The issue is not saving the item** it is in not-remembering to save the item.

It sounds to me like you have a problem remembering to save things, so you want a very technical solution implemented to resolve that.

**Comment by shiftyscales** [comment]
> I don't know of many use-cases for not packing and unpacking constantly?

I'd say the same use-cases otherwise highlighted elsewhere in this issue- I don't ever tend to pack everything up until such point as I'm done with whatever I'm editing / making, and I'm sure I'm not alone in that pattern.

> Similarly resonite could just dump a brson/resonitepackage/json into the data folder of the item as a way to recover?

As highlighted in my last post- Resonite already does this- if a world crashes in a managed way- it will attempt to dump the world state, and everything in it onto a local file on the user's computer.

**Comment by coolymike** [comment]
> I don't ever tend to pack everything up until such point as I'm done with whatever I'm editing / making, and I'm sure I'm not alone in that pattern.

"Prints" such as Redprint or OpenPrint parent the nodes to themselves. As soon as you've designated a slot for the nodes to be packed into, these prints ensure your nodes are in the same hierarchy as the rest of your project. "Creating a new project" would not have autosaves work properly, but editing an existing ProtoFlux project (with a print) would work just fine with autosaving a particular hierarchy.

Alongside that, many of the more complex projects use Dynamic Impulses and Dynamic Variables, which require the flux to be parented properly to function.

> Resonite already does this- if a world crashes in a managed way-

That's great, but most crashes are not managed, usually being Unity or Mono, not FrooxEngine "world crashes". I've personally lost hours of work over this.

**Comment by shiftyscales** [comment]
> "Prints" such as Redprint or OpenPrint parent the nodes to themselves.

Which is awesome- it's great when the community can produce and maintain tools they find useful and enhance their workflow. But I prefer working in 3D space and using the stock tooling as it's what I'm used to. I'm glad those improve the experience for you.

None of what you said negates my point that there are common scenarios in which a single "object" is not necessarily contained to a single hierarchy / slot. So having a mechanism that auto-saves a single slot is insufficient to resolve the underlying issue / request- particularly when part of the highlighted issue is unwanted materials being saved alongside the object (i.e. in the case of a full world save).

As we highlight in our [reporting requirements](https://github.com/Yellow-Dog-Man/Resonite-Issues#reporting-requirements)- it is best to focus on the high level issue first- rather than a specific solution. In this case, the underlying issue would be `"I sometimes lose work in progress due to forgetting or being unable to save it."`

It's best to leave the specifics of implementation to our coding team to work out and decide on- because requesting a specific implementation can often be unsuitable or otherwise have issues in its proposed design.

---

> That's great, but most crashes are not managed, usually being Unity or Mono, not FrooxEngine "world crashes". I've personally lost hours of work over this.

And I'm sorry about that, but in those cases then the better answer still would be to focus on resolving those underlying issues, i.e. eliminating crashes from the client in the first place wherever possible. As I've suggested to others, you may wish to see what else you can do in the meantime to improve your workflow to incorporate saving more often so you do not lose so much progress when those issues do arise.

Please keep on-topic, the underlying issues have already been well-established.
